addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/P1TelegramParser.java

```diff
@@ -0,0 +1,452 @@
+                    } else if (Character.isDigit(c)) {
+                        setState(State.DATA_OBIS_ID);
+                    } else if (c == '(') {
+                        setState(State.DATA_OBIS_VALUE);
+                    } else if (c == '!') {
+                        setState(State.CRC_VALUE);
+                    } else {
+                        handleUnexpectedCharacter(c);
+                        if (!lenientMode) {
+                            setState(State.WAIT_FOR_START);
+                        } // Other wise try to recover in lenient mode
+                    }
+                    break;
+
+                case CRC_VALUE:
+                    /*
+                     * Normally the P1 telegram ends with a \r\n sequence
+                     * If we already see a '/' character we also assume the current
+                     * P1 telegram is correctly finished
+                     */
+                    if (c == '\r' || c == '/') {
+                        logger.debug("telegramState {}, crcValue to check 0x{}", telegramState, crcValue);
+                        // Only perform CRC check if telegram is still ok
+                        if (telegramState == TelegramState.OK && crcValue.length() > 0) {
+                            if (Pattern.matches(CRC_PATTERN, crcValue)) {
+                                int crcP1Telegram = Integer.parseInt(crcValue.toString(), 16);
+                                int calculatedCRC = crc.getCurrentCRCCode();
+
+                                if (logger.isDebugEnabled()) {
+                                    logger.debug("received CRC value: {}, calculated CRC value: 0x{}", crcValue,
+                                            String.format("%04X", calculatedCRC));
+                                }
+                                if (crcP1Telegram != calculatedCRC) {
+                                    logger.debug("CRC value does not match, p1 Telegram failed");
+
+                                    telegramState = TelegramState.CRC_ERROR;
+                                }
+                            } else {
+                                telegramState = TelegramState.CRC_ERROR;
+                            }
+                        }
+
+                        if (factory != null && telegramListener != null) {
+                            /*
+                             * Send messages asynchronous to keep reading the Serial port.
+                             * If there are no messages to send, still let the listener
+                             * known that there are no messages and if the parsing was successful
+                             * This enables the listener to reinitialize the Serial Port for example
+                             */
+                            FutureTask<Void> task = new FutureTask<>(
+                                    new P1TelegramCallable(cosemObjects, telegramState));
+                            task.run();
```

Comment:
Not sure what is happening here. But this is a synchronize call. It will only be asynchroon if it's run by an executor service that starts this task in a new thread.

Reply:
Thank you for pointing this out. The executor service is missing indeed.